core/storage/pager.rs

```diff
@@ -1345,63 +1380,249 @@ impl Pager {
+                        "Freelist leaf page {} has overflow cells",
+                        leaf_page.get().id
+                    );
+                    leaf_page.get().contents.as_ref().unwrap().as_ptr().fill(0);
+                    let page_key = PageCacheKey::new(leaf_page.get().id);
+                    {
+                        let mut page_cache = self.page_cache.write();
+                        turso_assert!(
+                            page_cache.contains_key(&page_key),
+                            "page {} is not in cache",
+                            leaf_page.get().id
+                        );
+                    }
 
-        header_accessor::set_database_size(self, new_db_size)?;
+                    // Shift left all the other leaf pages in the trunk page and subtract 1 from the leaf count
+                    let remaining_leaves_count = (*number_of_freelist_leaves - 1) as usize;
+                    {
+                        let buf = page_contents.as_ptr();
+                        // use copy within the same page
+                        const LEAF_PTR_SIZE_BYTES: usize = 4;
+                        let offset_remaining_leaves_start =
+                            FREELIST_TRUNK_OFFSET_FIRST_LEAF + LEAF_PTR_SIZE_BYTES;
+                        let offset_remaining_leaves_end = offset_remaining_leaves_start
+                            + remaining_leaves_count * LEAF_PTR_SIZE_BYTES;
+                        buf.copy_within(
+                            offset_remaining_leaves_start..offset_remaining_leaves_end,
+                            FREELIST_TRUNK_OFFSET_FIRST_LEAF,
+                        );
+                    }
+                    // write the new leaf count
+                    page_contents.write_u32(
+                        FREELIST_TRUNK_OFFSET_LEAF_COUNT,
+                        remaining_leaves_count as u32,
+                    );
+                    self.add_dirty(trunk_page);
 
-        // FIXME: should reserve page cache entry before modifying the database
-        let page = allocate_page(new_db_size as usize, &self.buffer_pool, 0);
-        {
-            // setup page and add to cache
-            self.add_dirty(&page);
+                    header_accessor::set_freelist_pages(
+                        self,
+                        header_accessor::get_freelist_pages(self)? - 1,
+                    )?;
 
-            let page_key = PageCacheKey::new(page.get().id);
-            let mut cache = self.page_cache.write();
-            match cache.insert(page_key, page.clone()) {
-                Err(CacheError::Full) => Err(LimboError::CacheFull),
-                Err(_) => Err(LimboError::InternalError(
-                    "Unknown error inserting page to cache".into(),
-                )),
-                Ok(_) => Ok(page),
+                    *state = AllocatePageState::Start;
+                    return Ok(IOResult::Done(leaf_page));
```

Comment:
I think it's worth to add `*state = AllocatePageStart` in `reset_internal_states` now that pr is merged for safety.